in_toto/runlib.py

```diff
@@ -262,20 +263,68 @@ def execute_link(link_cmd_args, record_streams):
   # decide if s/he wants to see or store stdout/stderr
   # btw: we ignore them in the layout anyway
 
-  if record_streams:
-    # XXX: Use SpooledTemporaryFile if we expect very large outputs
-    stdout_file = tempfile.TemporaryFile()
-    stderr_file = tempfile.TemporaryFile()
-
-    return_value = subprocess.call(link_cmd_args,
-        stdout=stdout_file, stderr=stderr_file)
-
-    stdout_file.seek(0)
-    stderr_file.seek(0)
+  # Write subproc's stdout/err to a file and then read it back to python's output. 
+  # This simultaneously logs proc output while echoing it to the screen.
+  # Processes that demand ttys will still complain, but this maintians
+  # interactivity. Processes with buffered outputs or that use curses (eg nano)
+  # behave strangely. This should be portable but some processes (like Vim) 
+  # still break on Windows. 
 
-    stdout_str = stdout_file.read()
-    stderr_str = stderr_file.read()
+  if record_streams:
 
+    # Windows doesn't allow reopening TemporaryFile()s
+    stdout_file = tempfile.mkstemp()
+    stderr_file = tempfile.mkstemp()
+    stdout_name = stdout_file[1]
+    stderr_name = stderr_file[1]
+
+    # Open the files as r+b cause the subprocess to block indefinitely
+    with \
+    io.open(stdout_name, 'wb') as writer, \
+    io.open(stdout_name, 'rb', 1) as reader, \
+    io.open(stderr_name, 'wb') as err_write, \
+    io.open(stderr_name, 'rb', 1) as err_read:
```

Comment:
Note that you end up with 3 different filedescriptors per file here: `stdout_file[0] != writer.fileno() != reader.fileno()`.

I suppose you could use the filedescriptor that you get when calling `mkstemp()` for your reader or writer using `fdopen` instead of `open` then you wouldn't have to call `os.close` explicitly below (the `with` statement takes care of closing the files).

Reply:
Please use double indentation (4 spaces) for line continuation and don't break the line after the `with`, i.e.:
```python
with io.open(stdout_name, 'wb') as writer, \
    io.open(stdout_name, 'rb', 1) as reader,
    ...
 # Popen allows ...
```

Reply:
Fixed the indentation. 

Duplicating the file descriptors is intentional. I tried giving the subprocess the same fd as the main process and, while the output is logged to the file, the .read() in the main proc returns empty strings on each call. 

I can inline the temp creation into the with statement, but it looks kinda ugly and requires me to save the filenames for os.remove:
![image](https://user-images.githubusercontent.com/10876407/34019285-1aad5a62-e0fc-11e7-878c-eaa821df5cb5.png)